### PR TITLE
fix(nextly): enable localization reliably on existing collections, singles, and components

### PR DESCRIPTION
**Pull request by faisal-rx**
## Problem

Enabling Internationalization on an entity that already exists is broken in three distinct ways:

1. **Enable + add field in one save fails the apply (singles and components return 500).** Creating a single or component without localization, then adding a translatable field (e.g. `description`) and turning on Internationalization in the same save, crashes with `column "description" does not exist`. The enable migration seeds the new companion `_locales` table with `INSERT INTO companion (...) SELECT <all localized columns> FROM main`, but a field added in that same save was never on the main table, so the SELECT references a column that does not exist. On singles and components the whole apply fails; the entity is left half-migrated.

2. **camelCase field names strand data silently.** The seed and the main-table `DROP COLUMN` list were matched by raw field name, but physical columns are snake_cased by the storage descriptor (`subTitle` is stored as `sub_title`). The name never matched, so for any camelCase-named translatable field the enable plan contained no seed and no drop: the existing content stayed on a main column no read path looks at anymore (the value appears wiped in the UI), and the orphaned column became schema drift that the next apply flags as a destructive drop. No error is raised anywhere; the toggle reports success.

3. **Collections apply cleanly but the entry list 500s afterwards.** For collections the toggle goes through the metadata update path, which refreshed only the adapter's CRUD schema resolver. The entry-list read resolves its SELECT columns from the file manager's own schema cache, a separate cache that kept the pre-toggle table shape, so the list query kept selecting a column that had just moved to the companion: `column "description" does not exist` on every entry list load until restart.

### Seed and drop only columns that physically exist on main
`generate-up.ts` / `types.ts` / `reconcile-companion.ts`

`CompanionMigrationSpec` gains an optional `columnsOnMain`: the subset of localized columns that physically exist on the main table right now. The enable seed SELECTs and the main-table DROPs are restricted to that subset, while the companion CREATE still includes every localized column (a field added in the same save gets its companion column; it simply has no existing data to copy and nothing to drop). When nothing is on main and the entity has no Draft/Published status, the seed is skipped entirely rather than emitting an invalid empty INSERT.

The runtime toggle path populates `columnsOnMain` from the fields that existed before the save, resolving each name through the same column descriptor that produced the spec's columns, so the seed and drop always address the physical snake_case column (fix for problem 2 as well).

`columnsOnMain: undefined` means "all columns", which keeps the disable path byte-identical to before.

The file-migration path (`migrate:create`) gets the same treatment (review follow-up, `fbfd881e`): the enable plan populates `columnsOnMain` from the previous committed snapshot's main columns, so a code-first add-and-localize in one config change no longer generates a migration that fails on apply. The DOWN honors the same subset, re-adding and restoring exactly what the UP dropped, while still archiving every column's translations so rollback loses no language data.

### Refresh the read-path schema cache on a localization toggle
`collection-metadata-service.ts`

`registerRuntimeSchema` now calls `fileManager.refreshSchema(tableName, table)` with the freshly generated localized-aware runtime table, in addition to registering it on the adapter resolver. This mirrors exactly what the schema-apply path (`CollectionsHandler.refreshCollectionSchema`) already does; the metadata path was the only writer that skipped the read cache. Both services share the same `CollectionFileManager` instance, so the refresh reaches the exact cache the entry-list read consumes.

## Verification

Test-first: each seed bug was reproduced with a failing test before the fix (the add-in-same-save test failed with `no such column: "description"`, the camelCase test failed with `sub_title` seeded as `null`), and both pass after.

| Check | Result |
| --- | --- |
| i18n integration suite, SQLite | 20 files, 78 passed |
| i18n integration suite, Postgres | 20 files, 82 passed |
| i18n integration suite, MySQL | 20 files, 82 passed |
| Collections integration regression | 14 passed |
| `tsc --noEmit` (packages/nextly) | clean |

New coverage in `companion-transition.integration.test.ts` (runs the generated statements against a real database, not mocks):

- enable while a translatable field is added in the same save: companion gets both columns, the pre-existing value is seeded, the new field seeds as null, only the pre-existing column is dropped from main
- enable with a camelCase-named field: the value is copied under the physical `sub_title` column and the physical column is relocated off main

## Known limitations and deferrals

- **The read-cache fix has no automated test.** The `createTestNextly` harness resolves entry-list reads through a different path than production and never touches the file manager cache, so a test asserting this fix passes with or without it. Rather than ship a vacuous test, the fix is verified by code trace (both cache writers now match) and manual playground verification. Harness parity is a follow-up.

## Changeset

One changeset covering all published packages (patch, lockstep).

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed localized enablement when adding new translatable fields in the same save, ensuring companion creation and correct main-table column drops.
  - Improved companion migration planning to only seed and drop localized columns that already exist on the prior main table.
  - Updated runtime schema refresh so localized column changes are reflected immediately in collection entry-list reads.

- **Tests**
  - Expanded SQLite end-to-end coverage for companion transitions, including added columns, camelCase-to-snake_case mapping, and correct handling of fields without main-table equivalents.

### CODE REVIEW

File: packages/nextly/src/domains/i18n/migration/reconcile-companion.ts [hunk 1]
```diff
@@ -236,6 +236,16 @@ export function buildCompanionTransitionStatements(
...
+    // `wasLocalized` is false here, so every field that existed before this save lived on the
+    // main table; one added in this same save did not, and gets a companion column only. The
+    // names go through the same descriptor that produced `spec.columns`, so the seed/drop
+    // address the physical snake_case column (a field named `subTitle` is stored as
+    // `sub_title` — the raw field name would never match it).
+    const oldNames = new Set(oldFields.map(f => f.name));
+    spec.columnsOnMain = localizedNew
+      .filter(f => oldNames.has(f.name))
+      .map(f => fieldToLocalizedColumnSpec(f, dialect)?.name)
+      .filter((n): n is string => typeof n === "string");
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/i18n/migration/reconcile-companion.ts]: Confirmed and fixed in b1dd0667. Reproduced first with a failing test: a `component` field (no physical column) becoming a localized `text` in the same save produced `SELECT "gallery"`/`DROP COLUMN "gallery"` against a column main never had. `columnsOnMain` is now derived by resolving the OLD fields through the column descriptor and intersecting on physical column names, so a field whose previous type emitted no column (component, manyToMany) or a different column name is excluded from the seed and drop while still getting its companion column. Verified across SQLite, Postgres, and MySQL i18n suites.

File: packages/nextly/src/domains/collections/services/collection-metadata-service.ts [hunk 2]
```diff
@@ -242,6 +242,13 @@ export class CollectionMetadataService extends BaseService {
         // a restart — matching the just-created physical table.
         localized,
       });
+      // The entry-list read resolves its SELECT columns from the file manager's own schema
+      // cache, which is separate from the adapter resolver below. The schema-apply path
+      // refreshes both (CollectionsHandler.refreshCollectionSchema); this metadata path must
+      // too, or a localization toggle updates CRUD but leaves the read selecting the moved
+      // column from a table that no longer has it. Same shared file manager instance, so this
+      // reaches the exact cache the read consumes.
+      this.fileManager.refreshSchema(tableName, table);
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/collections/services/collection-metadata-service.ts:251]: Confirmed and fixed in b8ce9ed5. `loadCompanionSchema` caches the companion table under `<table>_locales` in the same file-manager registry, and this path refreshed only the main table there. The companion runtime table is now pushed into both consumers (file manager and adapter resolver), matching what the apply path does in `collection-dispatcher.ts`. The companion build was also hoisted out of the resolver guard so the read-cache refresh happens even when the resolver is absent.

File: .changeset/i18n-enable-add-field.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+"@nextlyhq/plugin-form-builder": patch
+"@nextlyhq/plugin-page-builder": patch
+"@nextlyhq/plugin-sdk": patch
+"@nextlyhq/prettier-config": patch
+"@nextlyhq/storage-s3": patch
+"@nextlyhq/storage-uploadthing": patch
+"@nextlyhq/storage-vercel-blob": patch
+"@nextlyhq/telemetry": patch
+"@nextlyhq/tsconfig": patch
+"@nextlyhq/ui": patch
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply .changeset/i18n-enable-add-field.md:20]: Confirmed and fixed in 7dcc5f74: `@nextlyhq/admin-css` added to the changeset. The package was published after this changeset was authored; recent changesets on main include it and this one now matches.

File: packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts [hunk 4]
```diff
@@ -86,6 +89,78 @@ describe("buildCompanionTransitionStatements — enable", () => {
...
+        name: string;
+      }[]
+    ).map(c => c.name);
+    expect(companionCols).toContain("heading");
+    expect(companionCols).toContain("description");
+
+    // The pre-existing value seeded; the brand-new field seeds as null (no source data).
+    const companionRow = sqlite
+      .prepare(`SELECT * FROM "single_hero_locales"`)
+      .get() as { heading: string; description: string | null };
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts:149]: Not reproducible against the actual file. Lines 124-126 and 155-157 each contain exactly one `.get()` call in a single multi-line chained expression (`sqlite.prepare(...).get() as {...}`); there is no duplicated statement. The file compiles (`tsc --noEmit` clean) and the suite passes 5/5 on SQLite, Postgres, and MySQL, which a parse failure would make impossible. Resolving as a false positive.

File: packages/nextly/src/domains/collections/services/collection-metadata-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/collections/services/collection-metadata-service.ts:251]: Agreed the gap is real, deferring the coverage. The `createTestNextly` harness resolves entry-list reads through a different path than production and never touches the file-manager cache, so a test written today passes with or without this call; a vacuous test would protect nothing while looking like coverage. The right fix is harness parity with the production read path (the same gap noted in the PR body for the main-table refresh), tracked as follow-up. Leaving this thread open as the tracker.

**Comment by mobeenabdullah** [comment]
@codex review pr

File: packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts [hunk 5]
```diff
@@ -86,6 +89,106 @@ describe("buildCompanionTransitionStatements — enable", () => {
...
+      oldFields: [...FIELDS, { name: "gallery", type: "component" as const }],
+      newFields: [...FIELDS, { name: "gallery", type: "text" as const }],
+      companionExists: false,
+    });
+    run(plan.statements);
+
+    // The companion has the column with no seeded data; the pre-existing field still seeded.
+    const companionRow = sqlite
+      .prepare(`SELECT * FROM "single_hero_locales"`)
+      .get() as { heading: string; gallery: string | null };
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts:211]: Ran your own verification script against the pushed commit: `sed -n '183,189p'` shows exactly one `.get()` call in a single chained expression (line 187), and `tsc --noEmit` exits clean. The suite also passes 5/5 on SQLite, Postgres, and MySQL, which a parse failure would make impossible. Same false positive as the previous thread on this file; the multi-line `sqlite.prepare(...).get() as {...}` chain appears to be getting double-counted. Resolving.

File: packages/nextly/src/domains/i18n/migration/generate-up.ts [hunk 6]
```diff
@@ -69,25 +69,43 @@ export function buildLocalizationUpStatements(
...
   const create = buildCompanionCreateStatement(spec);
 
+  // Only columns already on the main table can be seeded from or dropped. A field added and
+  // localized in the same save is in `columns` (so the companion gets it) but not on main, so
+  // it is excluded from the SELECT and the DROP. Undefined `columnsOnMain` means "all" — the
+  // file-migration path, where every localized column pre-exists on main.
+  const onMainSet = spec.columnsOnMain && new Set(spec.columnsOnMain);
+  const onMain = onMainSet
+    ? columns.filter(c => onMainSet.has(c.name))
+    : columns;
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/i18n/migration/generate-up.ts:82]: Confirmed and fixed in fbfd881e. You were right that the deferral rationale in the PR body did not hold: `prevMainColumnNames` from the previous committed snapshot is already threaded into `planCompanionMigration`, so the old-state data was available all along. The enable plan now sets `columnsOnMain` to the localized columns the previous main actually carried, so the seed SELECT and main-table DROP skip a field added and localized in the same config change while the companion CREATE still includes it. `buildLocalizationDownStatements` honors the same subset (re-adds and restores exactly what the UP dropped) while still archiving every column's translations so no language data is lost on rollback. The disable path passes `columnsOnMain` undefined and is byte-identical to before. Reproduced with a failing unit test first; plan units 9/9, migrate-create units 77/77, and i18n integration suites green on SQLite, Postgres, and MySQL.

File: packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts [hunk 7]
```diff
@@ -13,12 +13,15 @@ import { buildCompanionTransitionStatements } from "../reconcile-companion";
...
 let sqlite: Database.Database;
 
-/** A non-localized single's main table with a translatable `heading` column and a shared `views`. */
+/** A non-localized single's main table with a translatable `heading` column and a shared `views`.
+ *  `sub_title` is the physical column for a field NAMED `subTitle` — the storage descriptor
+ *  snake_cases field names, so the enable seed/drop must resolve columns the same way. */
 function createMainTable() {
   sqlite.exec(`CREATE TABLE "single_hero" (
     "id" TEXT PRIMARY KEY,
     "title" TEXT,
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/i18n/migration/__tests__/companion-transition.integration.test.ts:45]: Fixed in 5f1a50a9. The fixture no longer hand-writes the physical column: the field is declared once (`SUB_TITLE_FIELD`) and its column name AND DDL type are resolved through `fieldToLocalizedColumnSpec` + `ddlType`, the same helpers the migration generator uses, so the fixture tracks any future mapping change automatically. The INSERT and every assertion reference the derived name too. One correction on the stated mechanism: a descriptor rename would not have let the old test pass silently, because the generated seed would SELECT the new name against a fixture that lacks it and fail loudly. The real problem was the duplicated mapping knowledge the guideline exists to prevent, which this closes. Suite 5/5, tsc clean.

**Review by aqib-rx — APPROVED** [review]

**Comment by mobeenabdullah** [comment]
@codex @coderabbitai @greptile-apps review this PR
